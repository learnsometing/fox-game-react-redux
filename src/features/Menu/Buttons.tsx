import styled from 'styled-components';
import notSelected from './BUTTONS/NotSelected.png';
import hover from './BUTTONS/Hover.png';
import selected from './BUTTONS/Selected.png';

const Buttons = styled.div`
  position: absolute;
  top: 674px;
  left: 266px;
  height: 71px;
  width: 228px;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export default Buttons;

export const LeftButton = styled.div`
  background-image: url(${notSelected});
  background-position: 0px;
  width: 62px;
  height: 65px;
  overflow: hidden;
  background-repeat: no-repeat;
  background-color: transparent;
  border-color: transparent;
  cursor: pointer;

  &:hover,
  &:focus {
    background-image: url(${hover});
  }

  &:active {
    background-image: url(${selected});
  }
`;

export const MiddleButton = styled(LeftButton)`
  background-position: -84px 0px;
`;

export const RightButton = styled(LeftButton)`
  background-position: -167px 0px;
`;
