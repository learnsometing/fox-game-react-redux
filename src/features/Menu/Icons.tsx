import styled from 'styled-components';
import notSelected from './ICONS/NotSelected.png';
import highlighted from './ICONS/Hover.png';
import selected from './ICONS/Selected.png';

const Icons = styled.div`
  position: absolute;
  top: 572px;
  left: 194px;
  height: 67px;
  width: 374px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export default Icons;

interface IconProps {
  isHighlighted: boolean;
}

export const FishIcon = styled.div<IconProps>`
  background-image: ${(props) =>
    props.isHighlighted ? `url(${highlighted})` : `url(${notSelected})`};
  background-position: 0px;
  width: 56px;
  height: 65px;
  overflow: hidden;
  background-repeat: no-repeat;
  background-color: transparent;
  border-color: transparent;

  &:active {
    background-image: url(${selected});
  }
`;

export const PoopIcon = styled(FishIcon)`
  background-position: -162px 0px;
`;

export const WeatherIcon = styled(FishIcon)`
  background-position: -323px 0px;
`;
