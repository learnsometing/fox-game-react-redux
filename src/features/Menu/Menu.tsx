import React, { useState } from 'react';
import Buttons, { LeftButton, MiddleButton, RightButton } from './Buttons';
import Icons, { FishIcon, PoopIcon, WeatherIcon } from './Icons';

const Menu: React.FC = () => {
  const ICONS: string[] = ['fish-icon', 'poop-icon', 'weather-icon'];

  const [selectedIcon, setSelectedIcon] = useState(0);

  function selectPrev(): void {
    const selected = (2 + selectedIcon) % ICONS.length;
    setSelectedIcon(selected);
  }

  function selectNext(): void {
    const selected = (1 + selectedIcon) % ICONS.length;
    setSelectedIcon(selected);
  }

  function handleUserAction(): void {
    console.log(ICONS[selectedIcon]);
  }

  return (
    <>
      <Buttons>
        <LeftButton data-testid="left-btn" onClick={selectPrev} />
        <MiddleButton data-testid="middle-btn" onClick={handleUserAction} />
        <RightButton data-testid="right-btn" onClick={selectNext} />
      </Buttons>
      <Icons>
        <FishIcon isHighlighted={selectedIcon === 0} data-testid="fish-icon" />
        <PoopIcon isHighlighted={selectedIcon === 1} data-testid="poop-icon" />
        <WeatherIcon
          isHighlighted={selectedIcon === 2}
          data-testid="weather-icon"
        />
      </Icons>
    </>
  );
};

export default Menu;
