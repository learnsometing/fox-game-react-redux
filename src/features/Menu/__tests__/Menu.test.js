import React from 'react';
import Menu from '../Menu';
import notSelected from '../ICONS/NotSelected.png';
import highlighted from '../ICONS/Hover.png';
import { create } from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import 'jest-styled-components';

describe('Menu', () => {
  it('should match the snapshot', () => {
    const menu = create(<Menu />).toJSON();
    expect(menu).toMatchSnapshot();
  });

  it('clicking the left button should select the icon to the left', () => {
    const { queryByTestId } = render(<Menu />);
    const leftBtn = queryByTestId('left-btn');
    let weather = queryByTestId('weather-icon');
    expect(weather).toHaveStyleRule('background-image', `url(${notSelected})`);
    fireEvent.click(leftBtn);
    weather = queryByTestId('weather-icon');
    expect(weather).toHaveStyleRule('background-image', `url(${highlighted})`);
  });

  it('clicking the right button should select the icon to the right', () => {
    const { queryByTestId } = render(<Menu />);
    const rightBtn = queryByTestId('right-btn');
    let poop = queryByTestId('poop-icon');
    expect(poop).toHaveStyleRule('background-image', `url(${notSelected})`);
    fireEvent.click(rightBtn);
    poop = queryByTestId('poop-icon');
    expect(poop).toHaveStyleRule('background-image', `url(${highlighted})`);
  });

  it.todo('clicking the middle button should perform the selected action');
});
