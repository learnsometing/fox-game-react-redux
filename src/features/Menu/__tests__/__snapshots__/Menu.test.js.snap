// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should match the snapshot 1`] = `
Array [
  .c0 {
  position: absolute;
  top: 674px;
  left: 266px;
  height: 71px;
  width: 228px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  background-image: url(NotSelected.png);
  background-position: 0px;
  width: 62px;
  height: 65px;
  overflow: hidden;
  background-repeat: no-repeat;
  background-color: transparent;
  border-color: transparent;
  cursor: pointer;
}

.c1:hover,
.c1:focus {
  background-image: url(Hover.png);
}

.c1:active {
  background-image: url(Selected.png);
}

.c2 {
  background-image: url(NotSelected.png);
  background-position: 0px;
  width: 62px;
  height: 65px;
  overflow: hidden;
  background-repeat: no-repeat;
  background-color: transparent;
  border-color: transparent;
  cursor: pointer;
  background-position: -84px 0px;
}

.c2:hover,
.c2:focus {
  background-image: url(Hover.png);
}

.c2:active {
  background-image: url(Selected.png);
}

.c3 {
  background-image: url(NotSelected.png);
  background-position: 0px;
  width: 62px;
  height: 65px;
  overflow: hidden;
  background-repeat: no-repeat;
  background-color: transparent;
  border-color: transparent;
  cursor: pointer;
  background-position: -167px 0px;
}

.c3:hover,
.c3:focus {
  background-image: url(Hover.png);
}

.c3:active {
  background-image: url(Selected.png);
}

<div
    className="c0"
  >
    <div
      className="c1"
      data-testid="left-btn"
      onClick={[Function]}
    />
    <div
      className="c2"
      data-testid="middle-btn"
      onClick={[Function]}
    />
    <div
      className="c3"
      data-testid="right-btn"
      onClick={[Function]}
    />
  </div>,
  .c0 {
  position: absolute;
  top: 572px;
  left: 194px;
  height: 67px;
  width: 374px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  background-image: url(Hover.png);
  background-position: 0px;
  width: 56px;
  height: 65px;
  overflow: hidden;
  background-repeat: no-repeat;
  background-color: transparent;
  border-color: transparent;
}

.c1:active {
  background-image: url(Selected.png);
}

.c2 {
  background-image: url(NotSelected.png);
  background-position: 0px;
  width: 56px;
  height: 65px;
  overflow: hidden;
  background-repeat: no-repeat;
  background-color: transparent;
  border-color: transparent;
  background-position: -162px 0px;
}

.c2:active {
  background-image: url(Selected.png);
}

.c3 {
  background-image: url(NotSelected.png);
  background-position: 0px;
  width: 56px;
  height: 65px;
  overflow: hidden;
  background-repeat: no-repeat;
  background-color: transparent;
  border-color: transparent;
  background-position: -323px 0px;
}

.c3:active {
  background-image: url(Selected.png);
}

<div
    className="c0"
  >
    <div
      className="c1"
      data-testid="fish-icon"
    />
    <div
      className="c2"
      data-testid="poop-icon"
    />
    <div
      className="c3"
      data-testid="weather-icon"
    />
  </div>,
]
`;
